### ISSUE TITLE
Symbol not found: _getentropy

### ISSUE THREAD

**Issue by angkasapuri** [certbot/certbot#4067]
I'm trying to use certbot (with Let's encrypt) on my Mac Mini with the Server App. Installation with "brew install certbot" was successful, but I can't link it to apache with "certbot --apache". The error message is:

ImportError: dlopen(/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/cryptography/hazmat/bindings/_openssl.so, 2): Symbol not found: _getentropy
(full error see blow)

Does someone know what's going wrong? "certbot --apache certonly" fails with the same error.
I'm using Mac OS 10.11 El Capitan, with the Server App.

"apachectl -v" returns:
Server version: Apache/2.4.18 (Unix)
Server built:   Feb 20 2016 20:03:19

"brew -v" returns:
Homebrew 1.1.7-23-gc6d7c85
Homebrew/homebrew-core (git revision f5fa2; last commit 2017-01-15)

"brew list --versions"
augeas 1.7.0
certbot 0.10.1
dialog 1.3-20160209
gdbm 1.12
libxml2 2.9.4_2
mongodb 3.2.4 3.4.1
openssl 1.0.2j
openssl@1.1 1.1.0c
pkg-config 0.29.1_2
python 2.7.13
readline 7.0.1
redis 3.0.7 3.2.6
sqlite 3.16.2

I've run "brew update" and "brew upgrade" before (maybe that was a bad idea?). The commands are executed from an admin account (not root).

Full error message:
certbot --apache certonly
Traceback (most recent call last):
  File "/usr/local/bin/certbot", line 11, in <module>
    load_entry_point('certbot==0.10.1', 'console_scripts', 'certbot')()
  File "/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/pkg_resources/__init__.py", line 561, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2631, in load_entry_point
    return ep.load()
  File "/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2291, in load
    return self.resolve()
  File "/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2297, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/certbot/main.py", line 13, in <module>
    from acme import jose
  File "/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/acme/jose/__init__.py", line 37, in <module>
    from acme.jose.interfaces import JSONDeSerializable
  File "/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/acme/jose/interfaces.py", line 9, in <module>
    from acme.jose import util
  File "/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/acme/jose/util.py", line 5, in <module>
    import OpenSSL
  File "/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import rand, crypto, SSL
  File "/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/OpenSSL/rand.py", line 12, in <module>
    from OpenSSL._util import (
  File "/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/OpenSSL/_util.py", line 6, in <module>
    from cryptography.hazmat.bindings.openssl.binding import Binding
  File "/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 14, in <module>
    from cryptography.hazmat.bindings._openssl import ffi, lib
ImportError: dlopen(/usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/cryptography/hazmat/bindings/_openssl.so, 2): Symbol not found: _getentropy
  Referenced from: /usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/cryptography/hazmat/bindings/_openssl.so
  Expected in: /usr/lib/libSystem.B.dylib
 in /usr/local/Cellar/certbot/0.10.1/libexec/lib/python2.7/site-packages/cryptography/hazmat/bindings/_openssl.so

**Comment by bmw** [comment]
Any thoughts @DomT4?

**Comment by DomT4** [comment]
It's [this issue](https://github.com/pyca/cryptography/issues/3332). Can probably be worked around on Homebrew's side. I left the project for... reasons, a while back, but I'll take a look.

**Comment by bmw** [comment]
I'm sorry to hear you left the project, but I appreciate you taking a look. Do you know who I should reach out to in the future about potential issues with Certbot's Homebrew formula?

**Comment by DomT4** [comment]
@ilovezfs is the point person for `homebrew/core` these days. Generally if there's a problem he'll be able to offer some advice.

I'm still floating around GitHub & occasionally offer small Homebrew contributions so feel free to ping me as desired, but obviously, bear in mind my advice to them without the power of project maintainership behind it since October has little sway.

**Comment by angkasapuri** [comment]
@DomT4 thanks for pointing out [issue 3332](https://github.com/pyca/cryptography/issues/3332). Do I understand correctly that a workaround would be to uninstall Xcode and install the command line tools? This is a viable option for me.

**Comment by ilovezfs** [comment]
The cryptography package needs the same patch as it has in other formulae.

**Comment by ilovezfs** [comment]
e.g., https://github.com/Homebrew/homebrew-core/pull/8721#issuecomment-272825141

**Comment by DomT4** [comment]
@ilovezfs Aye. Already applied locally, just hadn't pushed yet. You have direct access so feel free to beat me to it.

**Comment by ilovezfs** [comment]
probably should create a cryptography formula since this doesn't appear to be getting fixed by upstream

**Comment by DomT4** [comment]
> thanks for pointing out issue 3332. Do I understand correctly that a workaround would be to uninstall Xcode and install the command line tools? This is a viable option for me.

I'm unsure, to be honest. I haven't looked at 10.11 since Sierra came out due to the complete development hell it became 😞. I _think_ it could work if you did that and possibly then reinstalled `certbot` with the `-s` flag, but I can't promise anything given my lack of testing. All Homebrew bottles are created on machines with both CLT and Xcode installed, so presumably that flaw is "built-in" to the El Capitan bottle.

**Comment by ilovezfs** [comment]
Maybe we can just require Sierra, else refuse to do anything.

**Comment by DomT4** [comment]
> probably should create a cryptography formula since this doesn't appear to be getting fixed by upstream

Might be worth nudging them. They're usually pretty quick at jumping on issues; the delay here is surprising.

> Maybe we can just require Sierra, else refuse to do anything.

I'm sure if you promised to immediately remove all the `.app`s from Homebrew regardless of usage as a bribe you might get that deal through _the team_.

**Comment by ilovezfs** [comment]
@angkasapuri why not upgrade to 10.12?

**Comment by ilovezfs** [comment]
@DomT4 Yes _we_ would likely go for that bargain.

**Comment by angkasapuri** [comment]
@ilovezfs actually I planned to upgrade to Sierra but was afraid the issues would only increase. But in this case I'll go ahead with it. Your time is too precious to be wasted on supporting old systems.

**Comment by ilovezfs** [comment]
@angkasapuri :+1: to upgrading to 10.12. There are very few issues left: https://github.com/Homebrew/homebrew-core/issues/6408

**Comment by DomT4** [comment]
@ilovezfs Did you want the PR or were you working on a formula for `cryptography`, presuming you can sneak it past Tim's dislike of Python libraries in the core? 😸

**Comment by ilovezfs** [comment]
A PR would be great.

**Comment by angkasapuri** [comment]
@ilovezfs duh, Sierra doesn't support my good old Mac Mini (2009)!
But your hint about removing Xcode and installing the command line tools worked, I managed to run certbot --apache.

However I'm now facing new issues, probably related to the fact that I'm using dyndns. I'm trying to figure out what to do now. If anyone has a pointer that would be great, but I also accept if this is now out of scope.

---

Saving debug log to /var/log/letsencrypt/letsencrypt.log
No names were found in your configuration files. Please enter in your domain
name(s) (comma and/or space separated)  (Enter 'c' to cancel):<myname>.dyndns.org
Obtaining a new certificate
Performing the following challenges:
tls-sni-01 challenge for <myname>.dyndns.org
No vhost exists with servername or alias of: <myname>.dyndns.org (or it's in a file with multiple vhosts, which Certbot can't parse yet). No vhost was selected. Please specify ServerName or ServerAlias in the Apache config, or split vhosts into separate files.
Falling back to default vhost *:443...
Waiting for verification...
Cleaning up challenges
Failed authorization procedure. <myname>.dyndns.org (tls-sni-01): urn:acme:error:unauthorized :: The client lacks sufficient authorization :: Incorrect validation certificate for TLS-SNI-01 challenge. Requested <number>.acme.invalid from <IP>:443. Received certificate containing ''

IMPORTANT NOTES:
 - The following errors were reported by the server:

   Domain: <myname>.dyndns.org
   Type:   unauthorized
   Detail: Incorrect validation certificate for TLS-SNI-01 challenge.
   Requested
   <number>.acme.invalid
   from <IP>.151:443. Received certificate containing ''

   To fix these errors, please make sure that your domain name was
   entered correctly and the DNS A record(s) for that domain
   contain(s) the right IP address.

**Comment by bmw** [comment]
This is definitely a different issue, but feel free to open a new issue for it if you're still having trouble. First things to check are shown in the output there:
```
To fix these errors, please make sure that your domain name was
entered correctly and the DNS A record(s) for that domain
contain(s) the right IP address.
```

**Comment by DomT4** [comment]
This was fixed in the 1.8 `cryptography` release, so shouldn't be an issue any more given `certbot` doesn't specify a specific `cryptography` version to be used. Was also fixed in Homebrew. Likely safe to close here.

**Comment by alex** [comment]
Our next release uses weak linkage to ensure that no matter what macOS version a wheel is built on, it will work fine on any supported macOS. This can be closed.
